File: win/directx/sdk-9.0/Samples/C#/Direct3D/Meshes/EnhancedMesh/EnhancedMesh.cs
//-----------------------------------------------------------------------------
// File: EnhancedMesh.cs
//
// Desc: Sample showing enhanced meshes in D3D
//
// Copyright (c) Microsoft Corporation. All rights reserved.
//-----------------------------------------------------------------------------
using System;
using System.Windows.Forms;
using Microsoft.DirectX;
using Microsoft.DirectX.Direct3D;
using Direct3D = Microsoft.DirectX.Direct3D;




namespace EnhancedMeshSample
{
	/// <summary>
	/// Application class. The base class (GraphicsSample) provides the generic 
	/// functionality needed in all Direct3D samples. MyGraphicsSample adds 
	/// functionality specific to this sample program.
	/// </summary>
	public class MyGraphicsSample : GraphicsSample
	{
		// Menu items specific for this application
		private MenuItem mnuOpenMesh = null;
		private MenuItem mnuMeshBreak = null;

		private GraphicsFont drawingFont = null;                 // Font for drawing text
		string initialDirectory = null;
		string meshFilename = "tiger.x"; // Filename of mesh
		
		Mesh systemMemoryMesh = null; // system memory version of mesh, lives through resize's
		Mesh enhancedMesh = null; // vid mem version of mesh that is enhanced
		int numberSegments; // number of segments per edge (tesselation level)
		Direct3D.Material[] meshMaterials = null; // array of materials
		Texture[] meshTextures = null;        // Array of textures, entries are NULL if no texture specified

		GraphicsArcBall arcBall = null;              // Mouse rotation utility
		Vector3 objectCenter;        // Center of bounding sphere of object
		float objectRadius = 0.0f;        // Radius of bounding sphere of object

		GraphicsStream adjacency = null; // Contains the adjacency info loaded with the mesh
		bool isUsingHardwareNPatches;
		bool isWireframe = false;


		

		/// <summary>
		/// Application constructor. Sets attributes for the app.
		/// </summary>
		public MyGraphicsSample()
		{
			// Set the window text
			this.Text = "Enhanced Mesh - N-Patches";
            try
            {
                // Load the icon from our resources
                System.Resources.ResourceManager resources = new System.Resources.ResourceManager(this.GetType());
                this.Icon = ((System.Drawing.Icon)(resources.GetObject("$this.Icon")));
            }
            catch
            {
                // It's no big deal if we can't load our icons, but try to load the embedded one
                try { this.Icon = new System.Drawing.Icon(this.GetType(), "directx.ico"); } 
                catch {}
            }

			arcBall = new GraphicsArcBall(this);
			initialDirectory = DXUtil.SdkMediaPath;
			drawingFont = new GraphicsFont("Arial", System.Drawing.FontStyle.Bold);
			enumerationSettings.AppUsesDepthBuffer = true;
			numberSegments = 2;

			// Add our new menu options
			this.mnuOpenMesh = new MenuItem();
			this.mnuMeshBreak = new MenuItem();

			// Add the Open File menu to the file menu
			this.mnuFile.MenuItems.Add(0, this.mnuMeshBreak);
			this.mnuMeshBreak.Text = "-";
			this.mnuFile.MenuItems.Add(0, this.mnuOpenMesh);
			this.mnuOpenMesh.Text = "Open File...";
			this.mnuOpenMesh.Shortcut = System.Windows.Forms.Shortcut.CtrlO;
			this.mnuOpenMesh.ShowShortcut = true;
			this.mnuOpenMesh.Click += new System.EventHandler(this.OpenMesh);

			// Set up our event handlers
			this.KeyDown += new System.Windows.Forms.KeyEventHandler(this.OnPrivateKeyDown);
		}





        /// <summary>
        /// The window has been created, but the device has not been created yet.  
        /// Here you can perform application-related initialization and cleanup that 
        /// does not depend on a device.
        /// </summary>
        protected override void OneTimeSceneInitialization()
		{
			// Set cursor to indicate that user can move the object with the mouse
			this.Cursor = System.Windows.Forms.Cursors.SizeAll;
		}




		/// <summary>
		/// Called once per frame, the call is the entry point for animating the scene.
		/// </summary>
		protected override void FrameMove()
		{
			// Setup world matrix
			Matrix matWorld = Matrix.Translation(-objectCenter.X,
				-objectCenter.Y,
				-objectCenter.Z);
			matWorld.Multiply(arcBall.RotationMatrix);
			matWorld.Multiply(arcBall.TranslationMatrix);
			device.Transform.World = matWorld;

			// Set up view matrix
			Vector3 vFrom= new Vector3(0, 0,-3*objectRadius);
			Vector3 vAt = new Vector3(0, 0, 0);
			Vector3 vUp = new Vector3(0, 1, 0);
			device.Transform.View = Matrix.LookAtLH(vFrom, vAt, vUp);
		}




		/// <summary>
		/// Called once per frame, the call is the entry point for 3d rendering. This 
		/// function sets up render states, clears the viewport, and renders the scene.
		/// </summary>
		protected override void Render()
		{
			// Clear the viewport
			device.Clear(ClearFlags.Target | ClearFlags.ZBuffer , System.Drawing.Color.Blue, 1.0f, 0);

			device.BeginScene();
			if (isUsingHardwareNPatches)
			{
				float fNumSegs;

				fNumSegs = (float)numberSegments;
				device.NPatchMode = fNumSegs;
			}

			// set and draw each of the materials in the mesh
			for (int i = 0; i < meshMaterials.Length; i++)
			{
				device.Material = meshMaterials[i];
				device.SetTexture(0, meshTextures[i]);

				enhancedMesh.DrawSubset(i);
			}

			if (isUsingHardwareNPatches)
			{
				device.NPatchMode = 0.0f;
			}

			// Output statistics
			drawingFont.DrawText(2,  1, System.Drawing.Color.Yellow, frameStats);
			drawingFont.DrawText(2, 20, System.Drawing.Color.Yellow, deviceStats);

			// Display info on mesh
			drawingFont.DrawText(2, 40, System.Drawing.Color.Yellow, "Number Segments: ");
			drawingFont.DrawText(150, 40, System.Drawing.Color.White, numberSegments.ToString());

			drawingFont.DrawText(2, 60, System.Drawing.Color.Yellow, "Number Faces: ");
			drawingFont.DrawText(150, 60, System.Drawing.Color.White, (enhancedMesh == null) ? "0" : enhancedMesh.NumberFaces.ToString());

			drawingFont.DrawText(2, 80, System.Drawing.Color.Yellow, "Number Vertices: ");
			drawingFont.DrawText(150, 80, System.Drawing.Color.White, (enhancedMesh == null) ? "0" : enhancedMesh.NumberVertices.ToString());

			device.EndScene();
		}




		/// <summary>
        /// The device has been created.  Resources that are not lost on
        /// Reset() can be created here -- resources in Pool.Managed,
        /// Pool.Scratch, or Pool.SystemMemory.  Image surfaces created via
        /// CreateImageSurface are never lost and can be created here.  Vertex
        /// shaders and pixel shaders can also be created here as they are not
        /// lost on Reset().
		/// </summary>
		protected override void InitializeDeviceObjects()
		{
			// Initialize the drawingFont's internal textures
			drawingFont.InitializeDeviceObjects(device);

			ExtendedMaterial[] mtrl = null;
			string path = DXUtil.FindMediaFile(initialDirectory, meshFilename);
			try
			{
				systemMemoryMesh = Mesh.FromFile(path, MeshFlags.SystemMemory, device, out adjacency, out mtrl);
			}
			catch
			{
				// Hide the error so we display a blue screen
				return;
			}

			// Lock the vertex buffer, to generate a simple bounding sphere
			VertexBuffer vb = null;
			try
			{
				vb = systemMemoryMesh.VertexBuffer;
				GraphicsStream vbStream = vb.Lock(0, 0, 0);
				objectRadius = Geometry.ComputeBoundingSphere(vbStream, systemMemoryMesh.NumberVertices, systemMemoryMesh.VertexFormat, out objectCenter);
			}
			finally
			{
				// Make sure we unlock the buffer if we fail
				if (vb != null)
					vb.Unlock();
			}

			meshMaterials = new Material[mtrl.Length];
			meshTextures = new Texture[mtrl.Length];
			for (int i = 0; i< mtrl.Length; i++)
			{
				meshMaterials[i] = mtrl[i].Material3D;
				if ((mtrl[i].TextureFilename != null) && ((mtrl[i].TextureFilename != string.Empty)))
					meshTextures[i] = TextureLoader.FromFile(device, DXUtil.FindMediaFile(null, mtrl[i].TextureFilename));
			}
			// Make sure there are normals, which are required for the tesselation
			// enhancement
			if ((systemMemoryMesh.VertexFormat & VertexFormats.Normal) != VertexFormats.Normal)
			{
				Mesh tempMesh = systemMemoryMesh.Clone(systemMemoryMesh.Options.Value, systemMemoryMesh.VertexFormat | VertexFormats.Normal, device);

				tempMesh.ComputeNormals();
				systemMemoryMesh.Dispose();
				systemMemoryMesh = tempMesh;
			}
		}




        /// <summary>
        /// Generates the enhanced mesh
        /// </summary>
		void GenerateEnhancedMesh(int newNumberSegs)
		{
			Mesh meshEnhancedSysMem = null;
			Mesh meshTemp = null;

			if (systemMemoryMesh == null)
				return;

			// if using hw, just copy the mesh
			if (isUsingHardwareNPatches)
			{
				meshTemp = systemMemoryMesh.Clone(MeshFlags.WriteOnly | MeshFlags.NPatches | (systemMemoryMesh.Options.Value & MeshFlags.Use32Bit), 
					systemMemoryMesh.VertexFormat, device);
			}
			else  // tesselate the mesh in sw
			{

				// Create an enhanced version of the mesh, will be in sysmem since source is
				try
				{
					meshEnhancedSysMem = Mesh.TessellateNPatches(systemMemoryMesh, adjacency, newNumberSegs, false);
				}
				catch
				{
					// If the tessellate failed, there might have been more triangles or vertices 
					// than can fit into a 16bit mesh, so try cloning to 32bit before tessellation

					meshTemp = systemMemoryMesh.Clone(MeshFlags.SystemMemory | MeshFlags.Use32Bit, systemMemoryMesh.VertexFormat, device);
					meshEnhancedSysMem = Mesh.TessellateNPatches(meshTemp, adjacency, newNumberSegs, false);
					meshTemp.Dispose();
				}

				// Make a video memory version of the mesh  
				// Only set WriteOnly if it doesn't use 32bit indices, because those 
				// often need to be emulated, which means that D3DX needs read-access.
				MeshFlags meshEnhancedFlags = meshEnhancedSysMem.Options.Value & MeshFlags.Use32Bit;
				if ((meshEnhancedFlags & MeshFlags.Use32Bit) != MeshFlags.Use32Bit)
					meshEnhancedFlags |= MeshFlags.WriteOnly;

				meshTemp = meshEnhancedSysMem.Clone(meshEnhancedFlags, systemMemoryMesh.VertexFormat, device);
			}

			enhancedMesh = meshTemp;
			numberSegments = newNumberSegs;
		}




		/// <summary>
        /// The device exists, but may have just been Reset().  Resources in
        /// Pool.Default and any other device state that persists during
        /// rendering should be set here.  Render states, matrices, textures,
        /// etc., that don't change during rendering can be set once here to
        /// avoid redundant state setting during Render() or FrameMove().
		/// </summary>
		protected override void RestoreDeviceObjects(System.Object sender, System.EventArgs e)
		{

			isUsingHardwareNPatches = Caps.DeviceCaps.SupportsNPatches; // Do we have hardware support?

			GenerateEnhancedMesh(numberSegments);

			// Setup render state
			renderState.Lighting = true;
			renderState.DitherEnable = true;
			renderState.ZBufferEnable = true;
			renderState.Ambient = System.Drawing.Color.FromArgb(0x33333333);
			sampleState[0].MagFilter = TextureFilter.Linear;
			sampleState[0].MinFilter = TextureFilter.Linear;

			// Setup the light
			GraphicsUtility.InitLight(device.Lights[0], LightType.Directional, 0.0f, -1.0f, 1.0f);
			device.Lights[0].Commit();
			device.Lights[0].Enabled = true;

			// Set the arcball parameters
			arcBall.SetWindow(device.PresentationParameters.BackBufferWidth, device.PresentationParameters.BackBufferHeight, 0.85f);
			arcBall.Radius = 1.0f;

			// Set the projection matrix
			float fAspect = device.PresentationParameters.BackBufferWidth / (float)device.PresentationParameters.BackBufferHeight;
			device.Transform.Projection = Matrix.PerspectiveFovLH((float)Math.PI/4, fAspect, objectRadius/64.0f,objectRadius*200.0f);

			if (isWireframe)
				device.RenderState.FillMode = FillMode.WireFrame;
			else
				device.RenderState.FillMode = FillMode.Solid;
		}


		
		
		/// <summary>
		/// Event Handler for windows messages
		/// </summary>
		private void OnPrivateKeyDown(object sender, System.Windows.Forms.KeyEventArgs e)
		{
			if (e.KeyCode == System.Windows.Forms.Keys.Up)
			{
				GenerateEnhancedMesh(numberSegments + 1);
			}
			if (e.KeyCode == System.Windows.Forms.Keys.Down)
			{
				GenerateEnhancedMesh(Math.Max(1, numberSegments - 1));
			}
			if (e.KeyCode == System.Windows.Forms.Keys.W)
			{
				device.RenderState.FillMode = FillMode.WireFrame;
				isWireframe = true;
			}
			if (e.KeyCode == System.Windows.Forms.Keys.S)
			{
				device.RenderState.FillMode = FillMode.Solid;
				isWireframe = false;
			}
		}




		/// <summary>
		/// Fired when a new mesh needs to be opened
		/// </summary>
		private void OpenMesh(object sender, EventArgs e)
		{
			// Display the OpenFileName dialog. Then, try to load the specified file
			System.Windows.Forms.OpenFileDialog ofn = new OpenFileDialog();
			ofn.Filter = ".X Files (*.x)|*.x";
			ofn.FileName = meshFilename;
			ofn.InitialDirectory = initialDirectory;
			ofn.Title = "Open Mesh File";
			ofn.CheckFileExists = true;
			ofn.Multiselect = false;
			ofn.ShowReadOnly = false;
			ofn.ShowDialog();
			System.IO.FileInfo fo = new System.IO.FileInfo(ofn.FileName);
			meshFilename = fo.Name;
			System.IO.Directory.SetCurrentDirectory(fo.DirectoryName);
			initialDirectory = fo.DirectoryName;

			// Destroy and recreate everything
			InvalidateDeviceObjects(null, null);
			DeleteDeviceObjects(null, null);
			try
			{
				InitializeDeviceObjects();
				RestoreDeviceObjects(null, null);
			}
			catch
			{
				MessageBox.Show("Error loading mesh: mesh may not\n" +
					"be valid. See debug output for\n" +
					"more information.\n\nPlease select " +
					"a different .x file.","EnhancedMesh", MessageBoxButtons.OK, MessageBoxIcon.Error);

			}
		}




        /// <summary>
        /// Called during device initialization, this code checks the device for 
        /// some minimum set of capabilities
        /// </summary>
		protected override bool ConfirmDevice(Caps caps, VertexProcessingType vertexProcessingType, 
			Format adapterFormat, Format backBufferFormat)
		{
			if ((vertexProcessingType == VertexProcessingType.PureHardware) && 
			    (!caps.DeviceCaps.SupportsNPatches) && (caps.DeviceCaps.SupportsRtPatches))
				return false;

			return true;
		}




        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        static void Main() 
        {
            using (MyGraphicsSample d3dApp = new MyGraphicsSample())
            {                                 
                if (d3dApp.CreateGraphicsSample())
                    d3dApp.Run();
            }
        }
	}
}

File: win/directx/sdk-9.0/Samples/VB.Net/Direct3D/Meshes/EnhancedMesh/EnhancedMesh.vb
'-----------------------------------------------------------------------------
' File: EnhancedMesh.vb
'
' Desc: Sample showing enhanced meshes in D3D
'
' Copyright (c) Microsoft Corporation. All rights reserved.
'-----------------------------------------------------------------------------
Imports System
Imports System.Windows.Forms
Imports Microsoft.DirectX
Imports Microsoft.DirectX.Direct3D
Imports Direct3D = Microsoft.DirectX.Direct3D


Namespace EnhancedMeshSample
    _
    '/ <summary>
    '/ Application class. The base class (GraphicsSample) provides the generic 
    '/ functionality needed in all Direct3D samples. MyGraphicsSample adds 
    '/ functionality specific to this sample program.
    '/ </summary>
    Public Class MyGraphicsSample
        Inherits GraphicsSample

        ' Menu items specific for this application
        Private mnuOpenMesh As MenuItem = Nothing
        Private mnuMeshBreak As MenuItem = Nothing

        Private drawingFont As GraphicsFont = Nothing ' Font for drawing text
        Private initialDirectory As String = Nothing
        Private meshFilename As String = "tiger.x" ' Filename of mesh
        Private systemMemoryMesh As Mesh = Nothing ' system memory version of mesh, lives through resize's
        Private enhancedMesh As Mesh = Nothing ' vid mem version of mesh that is enhanced
        Private numberSegments As Integer ' number of segments per edge (tesselation level)
        Private meshMaterials As Direct3D.Material() = Nothing ' array of materials
        Private meshTextures As Texture() = Nothing ' Array of textures, entries are NULL if no texture specified
        Private arcBall As GraphicsArcBall ' Mouse rotation utility
        Private objectCenter As Vector3 ' Center of bounding sphere of object
        Private objectRadius As Single = 0.0F ' Radius of bounding sphere of object
        Private adjacency As GraphicsStream = Nothing ' Contains the adjacency info loaded with the mesh
        Private isUsingHardwareNPatches As Boolean
        Private isWireframe As Boolean = False





        '/ <summary>
        '/ Application constructor. Sets attributes for the app.
        '/ </summary>
        Public Sub New()
            ' Set the window text
            Me.Text = "Enhanced Mesh - N-Patches"
            ' Load the icon from our resources
            'this.Icon = new System.Drawing.Icon( this.GetType(), "directx.ico" );
            arcBall = New GraphicsArcBall(Me)
            initialDirectory = DXUtil.SdkMediaPath
            drawingFont = New GraphicsFont("Arial", System.Drawing.FontStyle.Bold)
            enumerationSettings.AppUsesDepthBuffer = True
            numberSegments = 2

            ' Add our new menu options
            Me.mnuOpenMesh = New MenuItem()
            Me.mnuMeshBreak = New MenuItem()

            ' Add the Open File menu to the file menu
            Me.mnuFile.MenuItems.Add(0, Me.mnuMeshBreak)
            Me.mnuMeshBreak.Text = "-"
            Me.mnuFile.MenuItems.Add(0, Me.mnuOpenMesh)
            Me.mnuOpenMesh.Text = "Open File..."
            Me.mnuOpenMesh.Shortcut = System.Windows.Forms.Shortcut.CtrlO
            Me.mnuOpenMesh.ShowShortcut = True
            AddHandler Me.mnuOpenMesh.Click, AddressOf Me.OpenMesh

            ' Set up our event handlers
            AddHandler Me.KeyDown, AddressOf Me.OnPrivateKeyDown
        End Sub 'New






        '-----------------------------------------------------------------------------
        ' Name: OneTimeSceneInitialization()
        ' Desc: Called during initial app startup, this function performs all the
        '       permanent initialization.
        '-----------------------------------------------------------------------------
        Protected Overrides Sub OneTimeSceneInitialization()
            ' Set cursor to indicate that user can move the object with the mouse
            Me.Cursor = System.Windows.Forms.Cursors.SizeAll
        End Sub 'OneTimeSceneInitialization




        '/ <summary>
        '/ Called once per frame, the call is the entry point for animating the scene.
        '/ </summary>
        Protected Overrides Sub FrameMove()
            ' Setup world matrix
            Dim matWorld As Matrix = Matrix.Translation(-objectCenter.X, -objectCenter.Y, -objectCenter.Z)
            matWorld.Multiply(arcBall.RotationMatrix)
            matWorld.Multiply(arcBall.TranslationMatrix)
            device.Transform.World = matWorld

            ' Set up view matrix
            Dim vFrom As New Vector3(0, 0, -3 * objectRadius)
            Dim vAt As New Vector3(0, 0, 0)
            Dim vUp As New Vector3(0, 1, 0)
            device.Transform.View = Matrix.LookAtLH(vFrom, vAt, vUp)
        End Sub 'FrameMove




        '/ <summary>
        '/ Called once per frame, the call is the entry point for 3d rendering. This 
        '/ function sets up render states, clears the viewport, and renders the scene.
        '/ </summary>
        Protected Overrides Sub Render()
            ' Clear the viewport
            Device.Clear(ClearFlags.Target Or ClearFlags.ZBuffer, System.Drawing.Color.Blue, 1.0F, 0)

            Device.BeginScene()
            If isUsingHardwareNPatches Then
                Dim fNumSegs As Single

                fNumSegs = CSng(numberSegments)
                Device.NPatchMode = fNumSegs
            End If

            ' set and draw each of the materials in the mesh
            Dim i As Integer
            For i = 0 To meshMaterials.Length - 1
                Device.Material = meshMaterials(i)
                Device.SetTexture(0, meshTextures(i))

                enhancedMesh.DrawSubset(i)
            Next i

            If isUsingHardwareNPatches Then
                Device.NPatchMode = 0.0F
            End If

            ' Output statistics
            drawingFont.DrawText(2, 1, System.Drawing.Color.Yellow, frameStats)
            drawingFont.DrawText(2, 20, System.Drawing.Color.Yellow, deviceStats)

            ' Display info on mesh
            drawingFont.DrawText(2, 40, System.Drawing.Color.Yellow, "Number Segments: ")
            drawingFont.DrawText(150, 40, System.Drawing.Color.White, numberSegments.ToString())

            drawingFont.DrawText(2, 60, System.Drawing.Color.Yellow, "Number Faces: ")
            drawingFont.DrawText(150, 60, System.Drawing.Color.White, IIf(enhancedMesh Is Nothing, "0", enhancedMesh.NumberFaces.ToString()))

            drawingFont.DrawText(2, 80, System.Drawing.Color.Yellow, "Number Vertices: ")
            drawingFont.DrawText(150, 80, System.Drawing.Color.White, IIf(enhancedMesh Is Nothing, "0", enhancedMesh.NumberVertices.ToString()))

            Device.EndScene()
        End Sub 'Render





        '/ <summary>
        '/ Initialize scene objects.
        '/ </summary>
        Protected Overrides Sub InitializeDeviceObjects()
            ' Initialize the drawingFont's internal textures
            drawingFont.InitializeDeviceObjects(Device)

            Dim mtrl As ExtendedMaterial() = Nothing
            Dim sPath As String = DXUtil.FindMediaFile(initialDirectory, meshFilename)
            Try
                systemMemoryMesh = Mesh.FromFile(sPath, MeshFlags.SystemMemory, Device, adjacency, mtrl)
            Catch
                ' Hide the error so we display a blue screen
                Return
            End Try

            ' Lock the vertex buffer, to generate a simple bounding sphere
            Dim vb As VertexBuffer = Nothing
            Try
                vb = systemMemoryMesh.VertexBuffer
                Dim vbStream As GraphicsStream = vb.Lock(0, 0, 0)
                objectRadius = Geometry.ComputeBoundingSphere(vbStream, systemMemoryMesh.NumberVertices, systemMemoryMesh.VertexFormat, objectCenter)
            Finally
                ' Make sure we unlock the buffer if we fail
                If Not (vb Is Nothing) Then
                    vb.Unlock()
                End If
            End Try
            meshMaterials = New Material(mtrl.Length) {}
            meshTextures = New Texture(mtrl.Length) {}
            Dim i As Integer
            For i = 0 To mtrl.Length - 1
                meshMaterials(i) = mtrl(i).Material3D
                If Not (mtrl(i).TextureFilename Is Nothing) And mtrl(i).TextureFilename <> String.Empty Then
                    meshTextures(i) = TextureLoader.FromFile(Device, DXUtil.FindMediaFile(Nothing, mtrl(i).TextureFilename))
                End If
            Next i ' Make sure there are normals, which are required for the tesselation
            ' enhancement
            If (systemMemoryMesh.VertexFormat And VertexFormats.Normal) <> VertexFormats.Normal Then
                Dim tempMesh As Mesh = systemMemoryMesh.Clone(systemMemoryMesh.Options.Value, systemMemoryMesh.VertexFormat Or VertexFormats.Normal, Device)

                tempMesh.ComputeNormals()
                systemMemoryMesh.Dispose()
                systemMemoryMesh = tempMesh
            End If
        End Sub 'InitializeDeviceObjects





        '-----------------------------------------------------------------------------
        ' Name: GenerateEnhancedMesh()
        ' Desc: Generates the enhanced mesh
        '-----------------------------------------------------------------------------
        Sub GenerateEnhancedMesh(ByVal newNumberSegs As Integer)
            Dim meshEnhancedSysMem As Mesh = Nothing
            Dim meshTemp As Mesh = Nothing

            If systemMemoryMesh Is Nothing Then
                Return
            End If
            ' if using hw, just copy the mesh
            If isUsingHardwareNPatches Then
                meshTemp = systemMemoryMesh.Clone(MeshFlags.WriteOnly Or MeshFlags.NPatches Or (systemMemoryMesh.Options.Value And MeshFlags.Use32Bit), systemMemoryMesh.VertexFormat, Device)
                ' tesselate the mesh in sw
            Else

                ' Create an enhanced version of the mesh, will be in sysmem since source is
                Try
                    meshEnhancedSysMem = Mesh.TessellateNPatches(systemMemoryMesh, adjacency, newNumberSegs, False)
                Catch
                    ' If the tessellate failed, there might have been more triangles or vertices 
                    ' than can fit into a 16bit mesh, so try cloning to 32bit before tessellation
                    meshTemp = systemMemoryMesh.Clone(MeshFlags.SystemMemory Or MeshFlags.Use32Bit, systemMemoryMesh.VertexFormat, Device)
                    meshEnhancedSysMem = Mesh.TessellateNPatches(meshTemp, adjacency, newNumberSegs, False)
                    meshTemp.Dispose()
                End Try

                ' Make a video memory version of the mesh  
                ' Only set WriteOnly if it doesn't use 32bit indices, because those 
                ' often need to be emulated, which means that D3DX needs read-access.
                Dim meshEnhancedFlags As MeshFlags = meshEnhancedSysMem.Options.Value And MeshFlags.Use32Bit
                If (meshEnhancedFlags And MeshFlags.Use32Bit) <> MeshFlags.Use32Bit Then
                    meshEnhancedFlags = meshEnhancedFlags Or MeshFlags.WriteOnly
                End If
                meshTemp = meshEnhancedSysMem.Clone(meshEnhancedFlags, systemMemoryMesh.VertexFormat, Device)
            End If

            enhancedMesh = meshTemp
            numberSegments = newNumberSegs
        End Sub 'GenerateEnhancedMesh




        '/ <summary>
        '/ 
        '/ </summary>
        Protected Overrides Sub RestoreDeviceObjects(ByVal sender As System.Object, ByVal e As System.EventArgs)

            isUsingHardwareNPatches = Caps.DeviceCaps.SupportsNPatches ' Do we have hardware support?
            GenerateEnhancedMesh(numberSegments)

            ' Setup render state
            RenderState.Lighting = True
            RenderState.DitherEnable = True
            RenderState.ZBufferEnable = True
            RenderState.Ambient = System.Drawing.Color.FromArgb(&H33333333)
            sampleState(0).MagFilter = TextureFilter.Linear
            sampleState(0).MinFilter = TextureFilter.Linear

            ' Setup the light
            GraphicsUtility.InitLight(Device.Lights(0), LightType.Directional, 0.0F, -1.0F, 1.0F)
            Device.Lights(0).Commit()
            Device.Lights(0).Enabled = True

            ' Set the arcball parameters
            arcBall.SetWindow(Device.PresentationParameters.BackBufferWidth, Device.PresentationParameters.BackBufferHeight, 0.85F)
            arcBall.Radius = 1.0F

            ' Set the projection matrix
            Dim fAspect As Single = device.PresentationParameters.BackBufferWidth / CSng(device.PresentationParameters.BackBufferHeight)
            Device.Transform.Projection = Matrix.PerspectiveFovLH(CSng(Math.PI) / 4, fAspect, objectRadius / 64.0F, objectRadius * 200.0F)

            If isWireframe Then
                Device.RenderState.FillMode = FillMode.WireFrame
            Else
                Device.RenderState.FillMode = FillMode.Solid
            End If
        End Sub 'RestoreDeviceObjects




        '/ <summary>
        '/ Event Handler for windows messages
        '/ </summary>

        Private Sub OnPrivateKeyDown(ByVal sender As Object, ByVal e As System.Windows.Forms.KeyEventArgs)
            If e.KeyCode = System.Windows.Forms.Keys.Up Then
                GenerateEnhancedMesh((numberSegments + 1))
            End If
            If e.KeyCode = System.Windows.Forms.Keys.Down Then
                GenerateEnhancedMesh(Math.Max(1, numberSegments - 1))
            End If
            If e.KeyCode = System.Windows.Forms.Keys.W Then
                Device.RenderState.FillMode = FillMode.WireFrame
                isWireframe = True
            End If
            If e.KeyCode = System.Windows.Forms.Keys.S Then
                Device.RenderState.FillMode = FillMode.Solid
                isWireframe = False
            End If
        End Sub 'OnPrivateKeyDown




        '/ <summary>
        '/ Fired when a new mesh needs to be opened
        '/ </summary>
        Private Sub OpenMesh(ByVal sender As Object, ByVal e As EventArgs)
            ' Display the OpenFileName dialog. Then, try to load the specified file
            Dim ofn = New OpenFileDialog()
            ofn.Filter = ".X Files (*.x)|*.x"
            ofn.FileName = meshFilename
            ofn.InitialDirectory = initialDirectory
            ofn.Title = "Open Mesh File"
            ofn.CheckFileExists = True
            ofn.Multiselect = False
            ofn.ShowReadOnly = False
            ofn.ShowDialog()
            Dim fo As New System.IO.FileInfo(ofn.FileName)
            meshFilename = fo.Name
            System.IO.Directory.SetCurrentDirectory(fo.DirectoryName)
            initialDirectory = fo.DirectoryName

            ' Destroy and recreate everything
            InvalidateDeviceObjects(Nothing, Nothing)
            DeleteDeviceObjects(Nothing, Nothing)
            Try
                InitializeDeviceObjects()
                RestoreDeviceObjects(Nothing, Nothing)
            Catch
                MessageBox.Show("Error loading mesh: mesh may not" + ControlChars.Lf + "be valid. See debug output for" + ControlChars.Lf + "more information." + ControlChars.Lf + ControlChars.Lf + "Please select " + "a different .x file.", "EnhancedMesh", MessageBoxButtons.OK, MessageBoxIcon.Error)
            End Try
        End Sub 'OpenMesh





        '-----------------------------------------------------------------------------
        ' Name: ConfirmDevice()
        ' Desc: Called during device intialization, this code checks the device
        '       for some minimum set of capabilities
        '-----------------------------------------------------------------------------
        Protected Overrides Function ConfirmDevice(ByVal caps As Microsoft.DirectX.Direct3D.Caps, ByVal vertexProcessingType As VertexProcessingType, ByVal adapterFmt As Microsoft.DirectX.Direct3D.Format, ByVal backBufferFmt As Microsoft.DirectX.Direct3D.Format) As Boolean
            If vertexProcessingType = vertexProcessingType.PureHardware And Not caps.DeviceCaps.SupportsNPatches And caps.DeviceCaps.SupportsRtPatches Then
                Return False
            End If
            Return True
        End Function 'ConfirmDevice




        '/ <summary>
        '/ The main entry point for the application.
        '/ </summary>
        Shared Sub Main()
            Dim d3dApp As New MyGraphicsSample()
            If d3dApp.CreateGraphicsSample() Then
                d3dApp.Run()
            End If
        End Sub 'Main 
    End Class 'MyGraphicsSample 
End Namespace 'EnhancedMeshSample